### PR TITLE
RavenDB-13693 RavenDB should understand resource limits for burstable…

### PR DESCRIPTION
**Pull request by ayende**
… instances

### CODE REVIEW

File: src/Raven.Server/RavenServer.cs [hunk 1]
```diff
@@ -284,7 +301,119 @@ void ConfigureKestrel(KestrelServerOptions options)
...
+                }
+
+                return new DynamicJsonValue
+                {
+                    [nameof(Used)] = Used,
+                    [nameof(BaseCredits)] = BaseCredits,
+                    [nameof(MaxCredits)] = MaxCredits,
+                    [nameof(Threshold)] = Threshold,
+                    [nameof(RemainingCpuCredits)] = RemainingCpuCredits,
+                    [nameof(ThresholdReleasValue)] = ThresholdReleasValue,
```
Comment (ppekrol) [inline src/Raven.Server/RavenServer.cs]: typo: `ThresholdReleaseValue`

File: src/Raven.Server/RavenServer.cs [hunk 2]
```diff
@@ -284,7 +301,119 @@ void ConfigureKestrel(KestrelServerOptions options)
...
+                    [nameof(CreditsAccuredPerSecond)] = CreditsAccuredPerSecond,
```
Comment (ppekrol) [inline src/Raven.Server/RavenServer.cs]: typo: `CreditsAccuredPerSecond`... `CreditsOccuredPerSecond`? `CreditsAquiredPerSecond`? Not sure what you meant.

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Routing/RequestRouter.cs [hunk 3]
```diff
@@ -122,6 +122,15 @@ public async ValueTask HandlePath(RequestHandlerContext reqCtx)
...
                 if (reqCtx.Database != null)
                 {
+                    if (_ravenServer.CpuCreditsAlertRaised.IsRaised())
+                    {
+                        if (_ravenServer.Configuration.Server.CpuCreditsExhaustionShouldTriggerFailover)
+                        {
+                            RejectRequestBecauseOfCpuThreshold(context);
+                            return;
+                        }
+                    }
```
Comment (arekpalinski) [inline src/Raven.Server/Routing/RequestRouter.cs:132]: Maybe we can add an option to throttle requests then?

Comment (arekpalinski) [reply src/Raven.Server/Routing/RequestRouter.cs:132]: I mean when 'CpuCreditsExhaustionShouldTriggerFailover' is `false` then having additional option that will throttle the requests and reduce the load?

**Comment by ppekrol** [comment]
Couple of tests are failing here

File: src/Raven.Server/Routing/RequestRouter.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Routing/RequestRouter.cs:132]: That would just defer it, no? 
We can drop every Nth request, but that would probably suck

Comment (arekpalinski) [reply src/Raven.Server/Routing/RequestRouter.cs:132]: If we slow down the incoming load then we should be able to recover from current state earlier, don't you think? Once we manage to recover from it then slowly decrease the throttling (not immediately)

**Pull request merged**